### PR TITLE
Notebook fixes: Fix #4129, fix #4116, Fix #3913, fix empty results error 

### LINKED ISSUE #3913: Notebook: words selected in other cells should be unselected on cell change
This happens in Notebook cells where there's nothing selected, or often just for any function like `showEllipses()` where the `()` gets highlighted. Basically there are text selections like bracket matching that shouldn't be visible, but still are. 

image

### LINKED ISSUE #4116: Notebooks: Intellisense Doesn't Work using Add New Connection
Connected IntelliSense is not hooked up when selecting Add New Connection from the notebook toolbar.

If you create a new cell, it shows up; just doesn't show up if there is an existing cell, and you choose Add New Connection.

Repro:

Start ADS
Start new notebook from command palette (i.e. no connections show up in Attach to dropdown)
Select Add New Connection from Attach To dropdown, fill out info in connection dialog
Click inside existing code cell (do **not** click outside code cell first) [this means that the active cell id doesn't change), and type something like 'SELECT * FROM sys.'
Notice that the connected IntelliSense doesn't appear.

### LINKED ISSUE #4129: Overlapping command help windows in notebook
Issue Type: <b>Bug</b>

In the query editor, if you type a stored procedure parameters followed by a space, you get a helper window listing the parameters until you move on to another command or hit escape. This is pretty straightforward - you'd expect the help to follow you to additional lines in case your code style is to put the parameters on those lines, and if it's a hassle you can escape out.

help window 1

It seems to be watching for a syntactically correct command separator, as far as I can tell. In the following gif I am not putting "exec" before each proc (which is bad syntax and will result in a runtime error) and the window from the first proc just sticks around. 

help window 2

Unfortunately this looks very bad in a notebook, where the batch separation is implicit to the new code cell, not typed into the editor.

help window 3

Escape still works, but you have to hunt for the right cell to click on and hit escape if you get in this situation.

image

Importantly, the help does not go away after you've entered all the parameters, either (see last line of above image). So while in my examples I am just entering space for simplicity of reproduction, the answer isn't just "don't hit space" because you'll run into this if you enter all the parameters too, if you don't explicitly hit Esc.

Azure Data Studio version: azuredatastudio 1.5.0-insider (2dd71cbe2611189a433b6119bd10dad72ae6d104, 2019-02-20T03:25:50.936Z)
OS version: Windows_NT x64 10.0.17763

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-7600U CPU @ 2.80GHz (4 x 2904)|
|GPU Status|2d_canvas: enabled<br>checker_imaging: disabled_off<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>native_gpu_memory_buffers: disabled_software<br>rasterization: enabled<br>video_decode: enabled<br>video_encode: enabled<br>webgl: enabled<br>webgl2: enabled|
|Memory (System)|15.84GB (3.39GB free)|
|Process Argv|C:\Program Files\Azure Data Studio\azuredatastudio.exe|
|Screen Reader|no|
|VM|0%|

</details><details><summary>Extensions (8)</summary>

Extension|Author (truncated)|Version
---|---|---
vscode-sanddance|dan|0.0.1
newdatabase|kev|0.0.3
agent|Mic|0.36.1
import|Mic|0.6.0
profiler|Mic|0.6.0
server-report|Mic|0.1.4
sql-vnext|Mic|0.10.2
whoisactive|Mic|0.1.1

(1 theme extensions excluded)

</details>

### PR DESCRIPTION
**Pull request by kevcunnane**
- Fixes #4129 Overlapping command help windows in notebook
  - Do not show parameter hints for inactive cells, to avoid them hanging around when no longer selected
- Fixes #4116 Notebooks: Intellisense Doesn't Work using Add New Connection
  - Move connect/disconnect logic to 1 place (code component) instead of 2
  - Handle the case where you connect after choosing active cell. We now hook to the event and update connection
- Fix issues in sql session manager where result outputs 0 rows. This was failing to show the empty resultset contents, which is a regression vs. query editor. It also put unhandled error on the debug console
- Fix #3913 Notebook: words selected in other cells should be unselected on cell change

Note: after fix, now looks as follows. Need to do follow up to get correct grid min height

image

### CODE REVIEW

File: src/sql/parts/notebook/cellViews/code.component.ts [hunk 1]
```diff
@@ -135,6 +136,23 @@ export class CodeComponent extends AngularDisposable implements OnInit, OnChange
 		}
 	}
 
+	private updateConnectionState(isConnected: boolean) {
+		if (this.isSqlCodeCell()) {
+			let cellUri = this.cellModel.cellUri.toString();
+			if (!isConnected && this._model.notebookOptions.connectionService.isConnected(cellUri)) {
+				this._model.notebookOptions.connectionService.disconnect(cellUri).catch(e => console.log(e));
+			} else if (this._model.activeConnection && this._model.activeConnection.id !== '-1') {
```
Comment (rajmusuku) [inline src/sql/parts/notebook/cellViews/code.component.ts]: Just checking -  there is no scenario where connectionService is connected and connection id == -1 ? In that case we might need to add (this._model.activeConnection.id === '-1' || !isConnected) in If condition.

File: src/sql/parts/notebook/cellViews/code.component.ts [hunk 2]
```diff
@@ -135,6 +136,23 @@ export class CodeComponent extends AngularDisposable implements OnInit, OnChange
 		}
 	}
 
+	private updateConnectionState(isConnected: boolean) {
+		if (this.isSqlCodeCell()) {
+			let cellUri = this.cellModel.cellUri.toString();
+			if (!isConnected && this._model.notebookOptions.connectionService.isConnected(cellUri)) {
```
Comment (rajmusuku) [inline src/sql/parts/notebook/cellViews/code.component.ts]: this._model.notebookOptions.connectionService is always a valid object? shouldn't we guard it ?

File: src/sql/workbench/services/notebook/common/sqlSessionManager.ts [hunk 3]
```diff
@@ -398,32 +398,51 @@ export class SQLFuture extends Disposable implements FutureInternal {
 	public handleBatchEnd(batch: BatchSummary): void {
 		if (this.ioHandler) {
 			this.handleMessage(strings.format(elapsedTimeLabel, batch.executionElapsed));
+			this.processResultsSets(batch);
```
Comment (rajmusuku) [inline src/sql/workbench/services/notebook/common/sqlSessionManager.ts]: Isn't it  processResultSets?

File: src/sql/parts/notebook/cellViews/code.component.ts (hunk 1 above)
Comment (kevcunnane) [reply src/sql/parts/notebook/cellViews/code.component.ts]: Not that I know of. It starts at 0. We use this logic in other places so it seems a sensible way to identify invalid connections without special extra logic.

File: src/sql/parts/notebook/cellViews/code.component.ts (hunk 2 above)
Comment (kevcunnane) [reply src/sql/parts/notebook/cellViews/code.component.ts]: Definitely, will fix

File: src/sql/workbench/services/notebook/common/sqlSessionManager.ts (hunk 3 above)
Comment (kevcunnane) [reply src/sql/workbench/services/notebook/common/sqlSessionManager.ts]: Yes it is, will fix 😄

**Pull request merged**